### PR TITLE
feat: gateway endpoint discovery

### PR DESCRIPTION
**Pull request by JoanFM**
Goals:

When using endpoint, these are just used by WorkerRuntimes to skip processing Requests, but this can be used by gateway to avoid sending them, for this we need to do endpoint discovery

### CODE REVIEW

File: jina/proto/jina.proto [hunk 1]
```diff
@@ -175,3 +185,11 @@ service JinaRPC {
     }
 }
 
+/**
+ * jina gRPC service for DataRequests.
```
Comment (jacobowitz) [inline jina/proto/jina.proto:189]: That comment is not correct

File: jina/serve/runtimes/gateway/graph/topology_graph.py [hunk 2]
```diff
@@ -81,6 +85,12 @@ async def _wait_previous_and_send(
                             DataRequestHandler.reduce_requests(self.parts_to_send)
                         ]
 
+                    # avoid sending to executor which does not bind to this endpoint
+                    if endpoint is not None:
```
Comment (jacobowitz) [inline jina/serve/runtimes/gateway/graph/topology_graph.py:89]: why nesting and not concatenating with and?

File: jina/serve/runtimes/head/__init__.py [hunk 3]
```diff
@@ -251,6 +255,18 @@ async def process_control(self, request: ControlRequest, *args) -> ControlReques
             )
             raise
 
+    async def endpoint_discovery(self, empty, context) -> jina_pb2.EndpointsProto:
+        """
+        Process the received requests and return the result as a new request
```
Comment (jacobowitz) [inline jina/serve/runtimes/head/__init__.py:260]: that comment is not valid

File: jina/serve/runtimes/worker/__init__.py [hunk 4]
```diff
@@ -116,6 +120,19 @@ async def process_single_data(self, request: DataRequest, context) -> DataReques
         """
         return await self.process_data([request], context)
 
+    async def endpoint_discovery(self, empty, context) -> jina_pb2.EndpointsProto:
+        """
+        Process the received requests and return the result as a new request
```
Comment (jacobowitz) [inline jina/serve/runtimes/worker/__init__.py:125]: comment outdated

File: jina/serve/networking.py [hunk 5]
```diff
@@ -178,8 +179,34 @@ async def _init_stubs(self):
...
             self.stream_stub = stubs['jina.JinaRPC']
+            self.endpoints_discovery_stub = stubs['jina.JinaDiscoverEndpointsRPC']
             self._initialized = True
 
+        async def send_discover_endpoint(
+            self,
+            timeout: Optional[float] = None,
+        ) -> Tuple:
+            """
+            Send requests and uses the appropriate grpc stub for this
```
Comment (jacobowitz) [inline jina/serve/networking.py:190]: comment outdated

File: jina/serve/networking.py [hunk 6]
```diff
@@ -178,8 +179,34 @@ async def _init_stubs(self):
...
+
+            call_result = self.endpoints_discovery_stub.endpoint_discovery(
+                jina_pb2.google_dot_protobuf_dot_empty__pb2.Empty(),
+                timeout=timeout,
+            )
+            metadata, response = (
+                await call_result.trailing_metadata(),
+                await call_result,
+            )
+            return response, metadata
```
Comment (jacobowitz) [inline jina/serve/networking.py:208]:
```
return await self.endpoints_discovery_stub.endpoint_discovery(
                jina_pb2.google_dot_protobuf_dot_empty__pb2.Empty(),
                timeout=timeout)
```

you are not really interested in the metadata no?

File: jina/serve/networking.py [hunk 7]
```diff
@@ -503,6 +530,27 @@ def send_requests(
...
         return results
 
+    def send_discover_endpoint(
+        self,
+        deployment: str,
+        head: bool = True,
+        shard_id: Optional[int] = None,
+        timeout: Optional[float] = None,
+    ) -> asyncio.Task:
+        """Send a request to target via one or all of the pooled connections, depending on polling_type
```
Comment (jacobowitz) [inline jina/serve/networking.py:540]: comment not correct

File: jina/serve/networking.py [hunk 8]
```diff
@@ -651,17 +697,56 @@ async def task_wrapper(
...
+        # this wraps the awaitable object from grpc as a coroutine so it can be used as a task
+        # the grpc call function is not a coroutine but some _AioCall
+        async def task_wrapper():
+            for i in range(3):
+                try:
+                    return await connection.send_discover_endpoint(
+                        timeout=timeout,
+                    )
+                except AioRpcError as e:
+                    print(f' JOAN HERE ERROR {e}')
```
Comment (jacobowitz) [inline jina/serve/networking.py:719]: left over print

**Review by jacobowitz — CHANGES_REQUESTED** [review]

**Review by samsja — APPROVED** [review]
lgtm

**Pull request merged**